### PR TITLE
*: Support load data local

### PR DESCRIPTION
**Pull request by zimulala**
It only supports `load data local` and it doesn't supports enclosed and escaped in fields clause.
#1415 
#1543

### CODE REVIEW

File: executor/executor_test.go [hunk 1]
```diff
@@ -63,7 +66,7 @@ func (s *testSuite) SetUpSuite(c *C) {
 		c.Assert(err, IsNil)
 		s.store = store
 	}
-	log.SetLevelByString("warn")
+	log.SetLevelByString("info")
```
Comment (hanfei1991) [inline executor/executor_test.go]: why change log level?

File: plan/planbuilder.go [hunk 2]
```diff
@@ -510,6 +512,16 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
 	return insertPlan
 }
 
+func (b *planBuilder) buildLoadData(ld *ast.LoadDataStmt) Plan {
+	return &LoadData{
```
Comment (hanfei1991) [inline plan/planbuilder.go]: It's better to init an ID for it. You can refer to other logical plans.

File: server/conn.go [hunk 3]
```diff
@@ -270,6 +271,7 @@ func (cc *clientConn) dispatch(data []byte) error {
 	case mysql.ComQuit:
 		return io.EOF
 	case mysql.ComQuery:
+		log.Warnf("dispatch query")
```
Comment (hanfei1991) [inline server/conn.go]: a useless log ?

Comment (zimulala) [inline server/conn.go]: yep.

File: server/conn.go [hunk 4]
```diff
@@ -370,6 +372,73 @@ func (cc *clientConn) writeEOF() error {
...
+			return
+		}
+		err1 := cc.ctx.RollbackTxn()
+		if err1 != nil {
+			log.Errorf("load data rollback failed: %v", err1)
+		}
+	}()
+
+	if loadDataInfo == nil {
+		err = errors.Errorf("load data info is empty")
```
Comment (hanfei1991) [inline server/conn.go]: Using errors.new may be more clear. And you can combine it to next line.

File: server/conn.go [hunk 5]
```diff
@@ -370,6 +372,73 @@ func (cc *clientConn) writeEOF() error {
...
+		return errors.Trace(err)
+	}
+	err = cc.writeReq(loadDataInfo.Path)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var data1 []byte
```
Comment (hanfei1991) [inline server/conn.go]: I'm not quite understand the meanings of data1 and data2 :(

File: executor/executor_write.go [hunk 6]
```diff
@@ -407,6 +411,162 @@ func (e *DeleteExec) Close() error {
...
+	// TODO: support lines terminated is "".
+	if len(e.loadDataInfo.LinesInfo.Terminated) == 0 {
+		return nil, errors.New("don't support load data terminated is nil")
+	}
+
+	ctx := e.loadDataInfo.insertVal.ctx
+	val := ctx.Value(LoadDataVarKey)
+	if val != nil {
+		ctx.SetValue(LoadDataVarKey, nil)
+		return nil, errors.New("Load Data: previous load data option isn't close normal")
```
Comment (hanfei1991) [inline executor/executor_write.go]: close -> closed.

File: executor/executor_write.go [hunk 7]
```diff
@@ -407,6 +411,162 @@ func (e *DeleteExec) Close() error {
...
+
+// Next implements Executor Next interface.
+func (e *LoadData) Next() (*Row, error) {
+	// TODO: support load data without local field.
+	if !e.IsLocal {
+		return nil, errors.New("don't support load data without local field")
+	}
+	// TODO: support lines terminated is "".
+	if len(e.loadDataInfo.LinesInfo.Terminated) == 0 {
+		return nil, errors.New("don't support load data terminated is nil")
```
Comment (hanfei1991) [inline executor/executor_write.go]: don't -> Don't, same error appears in line 533.

File: executor/executor_write.go [hunk 8]
```diff
@@ -407,6 +411,162 @@ func (e *DeleteExec) Close() error {
...
+}
+
+// InsertData inserts data into specified table according to the specified format.
+// If it has the rest of data isn't completed the processing, then is returns without completed data.
+func (e *LoadDataInfo) InsertData(data1, data2 []byte) ([]byte, error) {
+	// TODO: support enclosed and escape.
+	var isNoEndData bool
+	var isLastData bool
+	var cols []string
+	if len(data1) > 0 && len(data2) <= 0 {
```
Comment (hanfei1991) [inline executor/executor_write.go]: Is it possible that len(data2) less than zero?

File: executor/executor_test.go [hunk 9]
```diff
@@ -371,6 +374,131 @@ func (s *testSuite) TestInsert(c *C) {
...
+func (s *testSuite) TestLoadData(c *C) {
+	defer testleak.AfterTest(c)()
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+
+	_, err := tk.Exec("load data local infile '/tmp/nonexistence.csv' into table load_data_test")
+	c.Assert(err, NotNil)
+	createSQL := `drop table if exists load_data_test;
+		create table load_data_test (id int PRIMARY KEY AUTO_INCREMENT, c1 int, c2 varchar(255) default "def", c3 int);`
+	tk.MustExec("commit")
```
Comment (hanfei1991) [inline executor/executor_test.go]: Why commit here ?

File: executor/executor_test.go [hunk 10]
```diff
@@ -371,6 +374,131 @@ func (s *testSuite) TestInsert(c *C) {
...
+	fields := &ast.FieldsClause{Terminated: "\t"}
+	lines := &ast.LinesClause{Starting: "", Terminated: "\n"}
+	ld := executor.NewLoadDataInfo(make([]types.Datum, 4), ctx, tbl)
+	ld.FieldsInfo = fields
+	ld.LinesInfo = lines
+
+	// data1 = nil, data2 = nil, fields and lines is default
+	_, err = ld.InsertData(nil, nil)
+	c.Assert(err, IsNil)
+	r := tk.MustQuery("select * from load_data_test;")
```
Comment (hanfei1991) [inline executor/executor_test.go]: How about put it after line 415 so as to use selectSQL?

**Comment by zimulala** [comment]
PTAL @hanfei1991 @coocood @shenli

File: executor/executor_write.go [hunk 11]
```diff
@@ -407,6 +411,162 @@ func (e *DeleteExec) Close() error {
...
+	Table      table.Table
+	FieldsInfo *ast.FieldsClause
+	LinesInfo  *ast.LinesClause
+}
+
+// InsertData inserts data into specified table according to the specified format.
+// If it has the rest of data isn't completed the processing, then is returns without completed data.
+func (e *LoadDataInfo) InsertData(data1, data2 []byte) ([]byte, error) {
+	// TODO: support enclosed and escape.
+	var isNoEndData bool
```
Comment (hanfei1991) [inline executor/executor_write.go]: Will "isNotTerminatedData" be better?

File: executor/executor_write.go [hunk 12]
```diff
@@ -407,6 +411,162 @@ func (e *DeleteExec) Close() error {
...
+
+	Path       string
+	Table      table.Table
+	FieldsInfo *ast.FieldsClause
+	LinesInfo  *ast.LinesClause
+}
+
+// InsertData inserts data into specified table according to the specified format.
+// If it has the rest of data isn't completed the processing, then is returns without completed data.
+func (e *LoadDataInfo) InsertData(data1, data2 []byte) ([]byte, error) {
```
Comment (hanfei1991) [inline executor/executor_write.go]: I think changing names of "data1/2" to remainedData and appendingData is better :D

File: executor/executor_write.go [hunk 13]
```diff
@@ -407,6 +411,162 @@ func (e *DeleteExec) Close() error {
...
+	var isLastData bool
+	var cols []string
+	if len(data1) > 0 && len(data2) == 0 {
+		isLastData = true
+		data1, data2 = data2, data1
+	}
+	startingLen := len(e.LinesInfo.Starting)
+	terminatedLen := len(e.LinesInfo.Terminated)
+	for len(data2) > 0 {
+		end := strings.Index(string(data2), e.LinesInfo.Terminated)
```
Comment (hanfei1991) [inline executor/executor_write.go]: endIdx will make more sense

File: executor/executor_test.go [hunk 14]
```diff
@@ -371,6 +374,130 @@ func (s *testSuite) TestInsert(c *C) {
...
+		{nil, []byte("\\2\\3\\4|!#^"), []string{}},
+
+		// data1 != nil, data2 = nil
+		{[]byte("xxx\\2\\3\\4"), nil, []string{fmt.Sprintf("%v %v %v %v", 18, 2, []byte("3"), 4)}},
+		{[]byte("\\2\\3\\4|!#^"), nil, []string{}},
+		{[]byte("\\2\\3\\4|!#^xxx18\\22\\33\\44|!#^"), nil, []string{fmt.Sprintf("%v %v %v %v", 18, 22, []byte("33"), 44)}},
+
+		// data1 != nil, data2 != nil
+		{[]byte("xxx\\2\\3"), []byte("\\4\\5|!#^"), []string{fmt.Sprintf("%v %v %v %v", 19, 2, []byte("3"), 4)}},
+		{[]byte("xxx\\2\\3"), []byte("4\\5|!#^"), []string{fmt.Sprintf("%v %v %v %v", 20, 2, []byte("34"), 5)}},
```
Comment (hanfei1991) [inline executor/executor_test.go]: add a test that starting flag and terminated flag cross the data1 and data2.

File: executor/executor_write.go [hunk 15]
```diff
@@ -407,6 +411,162 @@ func (e *DeleteExec) Close() error {
...
+				data2 = data2[end+terminatedLen:]
+				continue
+			}
+			data2 = data2[startingLen:]
+			end -= startingLen
+		}
+		if len(data1) > 0 {
+			data1 = append(data1, data2[:end]...)
+			if isNoEndData && !isLastData {
+				data2 = data1
```
Comment (hanfei1991) [inline executor/executor_write.go]: Add one more judgement to hold the case that a end flag cross two packets.

File: executor/executor_write.go [hunk 16]
```diff
@@ -407,6 +411,162 @@ func (e *DeleteExec) Close() error {
...
+			}
+			data1 = nil
+		} else {
+			cols = strings.Split(string(data2[:end]), e.FieldsInfo.Terminated)
+		}
+		if isNoEndData && !isLastData {
+			break
+		}
+		for i := 0; i < len(e.row); i++ {
+			if i > len(cols)-1 {
```
Comment (hanfei1991) [inline executor/executor_write.go]: i >= len(cols) will be more clear.

File: executor/executor_write.go [hunk 17]
```diff
@@ -407,6 +411,162 @@ func (e *DeleteExec) Close() error {
...
+				e.row[i].SetNull()
+				continue
+			}
```
Comment (hanfei1991) [inline executor/executor_write.go]: Will using if ... else ... statements be more clear?

File: executor/executor_write.go [hunk 18]
```diff
@@ -407,6 +411,162 @@ func (e *DeleteExec) Close() error {
...
+		row, err := e.insertVal.fillRowData(e.Table.Cols(), e.row, true)
+		if err != nil {
+			log.Warnf("Load Data: insert data:%v failed:%v", e.row, errors.ErrorStack(err))
+		}
+		_, err = e.Table.AddRecord(e.insertVal.ctx, row)
+		if err != nil {
+			log.Warnf("Load Data: insert data:%v failed:%v", row, errors.ErrorStack(err))
+		}
+		e.insertVal.currRow++
+		if len(data2) <= end+terminatedLen {
```
Comment (hanfei1991) [inline executor/executor_write.go]: When this condition will be true?

File: executor/executor_write.go (hunk 17 above)
Comment (zimulala) [inline executor/executor_write.go]: I don't think so.

Comment (hanfei1991) [inline executor/executor_write.go]: It all depends you :)

File: executor/executor_write.go [hunk 19]
```diff
@@ -407,6 +411,162 @@ func (e *DeleteExec) Close() error {
...
+	FieldsInfo *ast.FieldsClause
+	LinesInfo  *ast.LinesClause
+}
+
+// InsertData inserts data into specified table according to the specified format.
+// If it has the rest of data isn't completed the processing, then is returns without completed data.
+func (e *LoadDataInfo) InsertData(data1, data2 []byte) ([]byte, error) {
+	// TODO: support enclosed and escape.
+	var isNoEndData bool
+	var isLastData bool
```
Comment (coocood) [inline executor/executor_write.go]: It's confusing with `last data` and `end data`.

File: executor/executor_write.go (hunk 12 above)
Comment (coocood) [inline executor/executor_write.go]: how about `prevData` and `curData`?

**Comment by zimulala** [comment]
PTAL @hanfei1991 @coocood

File: executor/executor_write.go [hunk 20]
```diff
@@ -407,6 +411,180 @@ func (e *DeleteExec) Close() error {
...
+		prevData, curData = curData, prevData
+	}
+	for len(curData) > 0 {
+		endIdx := strings.Index(string(curData), e.LinesInfo.Terminated)
+		if endIdx == -1 {
+			if len(prevData) > 0 {
+				curData = append(prevData, curData...)
+			}
+			endIdx = len(curData)
+			notLineEnd = true
```
Comment (coocood) [inline executor/executor_write.go]: We can check if Terminated can be found after we have append prevData and curData, if not found, we can return the appended data directly, then we only need to handle the case when there is terminate string.

**Comment by zimulala** [comment]
PTAL @coocood @hanfei1991

File: executor/executor_write.go [hunk 21]
```diff
@@ -407,6 +411,251 @@ func (e *DeleteExec) Close() error {
...
+	idx = strings.Index(string(curData), e.LinesInfo.Starting)
+	if idx != -1 {
+		return nil, curData[idx:], startingLen
+	}
+
+	// no starting symbol
+	if len(curData) < startingLen {
+		restStart = curData
+	} else {
+		restStart = curData[len(curData)-startingLen+1:]
```
Comment (hanfei1991) [inline executor/executor_write.go]: Why assign it to restStart but return curData self?

File: executor/executor_write.go [hunk 22]
```diff
@@ -407,6 +411,251 @@ func (e *DeleteExec) Close() error {
...
+	}
+	return nil, restStart, -1
+}
+
+// getLine returns a line, curData, the next data start index and a bool value.
+// If it has starting symbol the bool is true, otherwise is false.
+func (e *LoadDataInfo) getLine(prevData, curData []byte) ([]byte, []byte, int, bool) {
+	prevData, curData, curStartIdx := e.getStartingInfo(prevData, curData)
```
Comment (hanfei1991) [inline executor/executor_write.go]: You can calculate curStartIdx outside

File: executor/executor_write.go [hunk 23]
```diff
@@ -407,6 +411,251 @@ func (e *DeleteExec) Close() error {
...
+		nextDataIdx        int
+		isEOF, hasStarting bool
+	)
+	cols := make([]string, 0, len(e.row))
+	if len(prevData) > 0 && len(curData) == 0 {
+		isEOF = true
+		prevData, curData = curData, prevData
+	}
+	for len(curData) > 0 {
+		line, curData, nextDataIdx, hasStarting = e.getLine(prevData, curData)
```
Comment (hanfei1991) [inline executor/executor_write.go]: You can truncate curData inside "getLine".

File: executor/executor_write.go [hunk 24]
```diff
@@ -407,6 +411,251 @@ func (e *DeleteExec) Close() error {
...
+	// terminated symbol in the curData
+	if len(prevData) == 0 {
+		nextDataIdx := curStartIdx + endIdx + terminatedLen
+		return curData[curStartIdx : curStartIdx+endIdx], curData, nextDataIdx, true
+	}
+
+	// terminated symbol in the curData
+	prevData = append(prevData, curData[:curStartIdx+endIdx+terminatedLen]...)
+	endIdx = strings.Index(string(prevData[startingLen:]), e.LinesInfo.Terminated)
+	lineLen := startingLen + endIdx + terminatedLen
```
Comment (hanfei1991) [inline executor/executor_write.go]: You can calculate lineLen before line 516.

File: executor/executor_write.go [hunk 25]
```diff
@@ -407,6 +411,251 @@ func (e *DeleteExec) Close() error {
...
+		endIdx = strings.Index(string(curData[curStartIdx:]), e.LinesInfo.Terminated)
+	}
+	if endIdx == -1 {
+		// no terminated symbol
+		if len(prevData) == 0 {
+			return nil, curData, 0, true
+		}
+
+		// terminated symbol in the middle of prevData and curData
+		curData = append(prevData, curData...)
```
Comment (hanfei1991) [inline executor/executor_write.go]: Maybe append curData[terminatedLen-1] is enough, and return rest curData

File: executor/executor_write.go (hunk 24 above)
Comment (zimulala) [inline executor/executor_write.go]: They are not the same value.

Comment (hanfei1991) [inline executor/executor_write.go]: ok, my fault.

File: executor/executor_write.go (hunk 23 above)
Comment (zimulala) [inline executor/executor_write.go]: Here can handle line549 - 553 only once.

**Comment by zimulala** [comment]
PTAL @hanfei1991

**Comment by hanfei1991** [comment]
LGTM :+1:

**Comment by hanfei1991** [comment]
@coocood @shenli PTAL

File: executor/executor_write.go [hunk 26]
```diff
@@ -407,6 +411,254 @@ func (e *DeleteExec) Close() error {
...
+	Path       string
+	Table      table.Table
+	FieldsInfo *ast.FieldsClause
+	LinesInfo  *ast.LinesClause
+}
+
+// getVaildData returns prevData and curData that starts from starting symbol.
+// If the data doesn't have starting symbol, prevData is nil and curData is curData[len(curData)-startingLen+1:].
+// If curData size less than startingLen, curData is returned directly.
+func (e *LoadDataInfo) getVaildData(prevData, curData []byte) ([]byte, []byte) {
```
Comment (coocood) [inline executor/executor_write.go]: Valid

File: executor/executor_write.go [hunk 27]
```diff
@@ -407,6 +411,254 @@ func (e *DeleteExec) Close() error {
...
+	startingLen := len(e.LinesInfo.Starting)
+	if startingLen == 0 {
+		return prevData, curData
+	}
+
+	// starting symbol in the prevData
```
Comment (coocood) [inline executor/executor_write.go]: wrap this line 443 to line 461 into `if len(prevData) > 0 { ... }` is more efficient.

File: executor/executor_write.go [hunk 28]
```diff
@@ -407,6 +411,254 @@ func (e *DeleteExec) Close() error {
...
+	idx := strings.Index(string(prevData), e.LinesInfo.Starting)
+	if idx != -1 {
+		return prevData[idx:], curData
+	}
+
+	// starting symbol in the middle of prevData and curData
+	restStart := make([]byte, 0, startingLen-1)
```
Comment (coocood) [inline executor/executor_write.go]: Why make a slice if you just assign it to other slices?

File: executor/executor_write.go [hunk 29]
```diff
@@ -407,6 +411,254 @@ func (e *DeleteExec) Close() error {
...
+		nextDataIdx := curStartIdx + endIdx + terminatedLen
+		return curData[curStartIdx : curStartIdx+endIdx], curData, nextDataIdx, true
+	}
+
+	// terminated symbol in the curData
+	prevData = append(prevData, curData[:curStartIdx+endIdx+terminatedLen]...)
+	endIdx = strings.Index(string(prevData[startingLen:]), e.LinesInfo.Terminated)
+	lineLen := startingLen + endIdx + terminatedLen
+	if endIdx >= prevLen {
+		return prevData[startingLen : startingLen+endIdx], prevData, lineLen, true
```
Comment (coocood) [inline executor/executor_write.go]: The returned prevData may lost the rest lines in curData?

**Comment by coocood** [comment]
LGTM

**Pull request merged**